lib/core/checker/type_expression_checker.cc

```diff
@@ -0,0 +1,222 @@
+#include "type_expression_checker.h"
+
+#include "core/ast/ast.h"
+#include "core/ast/expression.h"
+#include "location.hh"
+
+#include <string>
+#include <vector>
+
+namespace dbuf::checker {
+
+TypeExpressionChecker::TypeExpressionChecker(ast::AST &ast)
+    : ast_(ast) {}
+
+void TypeExpressionChecker::GetConstructorToEnum() {
+  for (const auto &ast_enum : ast_.enums) {
+    for (const auto &rule : ast_enum.second.pattern_mapping) {
+      for (const auto &constructor : rule.outputs) {
+        constructor_to_enum_[constructor.identifier.name] = ast_enum.second.identifier.name;
+      }
+    }
+  }
+}
+
+void TypeExpressionChecker::PushScope() {
+  scopes_.emplace_back();
+}
+
+void TypeExpressionChecker::PopScope() {
+  if (scopes_.empty()) {
+    throw std::logic_error("Can't delete scope from empty scopes.");
+  }
+  scopes_.pop_back();
+}
+
+bool TypeExpressionChecker::IsInScope(InternedString name) {
+  return std::ranges::any_of(scopes_.begin(), scopes_.end(), [&name](auto &scope) {
+    return scope.contains(name);
+  });
+}
+
+void TypeExpressionChecker::AddName(InternedString name, InternedString type) {
```

Comment:
What if the variable of type Vec n? Is the string for the type sufficient?

Reply:
To be honest this map is more like "variable to typename". I need it to look for message or enum I need to check